bindings/pyroot/pythonizations/python/ROOT/pythonization/__init__.py

```diff
@@ -7,3 +7,284 @@
+
+        return cppyy_pythonizor
+
+    return pythonization_impl
+
+def _parse_target(target):
+    '''
+    Helper function to check the type of the `target` argument specified by the
+    user in a @pythonization decorator.
+    It also returns the received fully-qualified class name(s)/prefix(es) and
+    the corresponding scope(s), removing any repetition.
+
+    Args:
+        target (string/iterable[string]): fully-qualified class name(s)/
+            prefix(es).
+
+    Returns:
+        tuple[set,set]: fully-qualified class name(s)/prefix(es) and scope(s)
+            in `target`, with no repetitions.
+    '''
+
+    scopes = set()
+    scope_to_targets = {}
+
+    if _is_string(target):
+        _register_target(target, scopes, scope_to_targets)
+    else:
+        try:
+            for name in target:
+                if _is_string(name):
+                    _register_target(name, scopes, scope_to_targets)
+                else:
+                    raise TypeError()
+        except TypeError:
+            raise TypeError('Invalid type of "target" argument in @pythonization: '
+                            'must be string or iterable of strings')
+
+    return scopes, scope_to_targets
+
+def _is_string(o):
+    '''
+    Checks if object is a string.
+
+    Args:
+        o (any type): object to be checked.
+
+    Returns:
+        True if object is a string, False otherwise.
+    '''
+
+    if sys.version_info >= (3, 0):
+        return isinstance(o, str)
+    else:
+        return isinstance(o, basestring)
+
+def _register_target(target, scopes, scope_to_targets):
+    '''
+    Registers all the targets of a pythonizor per scope.
+
+    Args:
+        target (string): fully-qualified class name/prefix.
+        scopes (list): list of scopes to add the scope of `target` to.
+        scope_to_targets (dict): map from scope to list of targets that have
+            been registered for that scope.
+    '''
+
+    scope, class_name = _split_scope_and_class(target)
+    scopes.add(scope)
+    targets = scope_to_targets.get(scope)
+    if targets is None:
+        targets = []
+        scope_to_targets[scope] = targets
+    targets.append(class_name)
+
+def _get_scope(fqn):
+    '''
+    Parses and returns the scope in `fqn`.
+    Example: if `fqn` is "NS1::NS2::C", "NS1::NS2" is returned.
+
+    Args:
+        fqn (string): fully-qualified class name.
+
+    Returns:
+        string: scope of `fqn`.
+    '''
+
+    pos = _find_namespace_end(fqn)
+    if pos < 0: # global namespace
+        return ''
+    else:
+        return fqn[:pos]
+
+def _split_scope_and_class(fqn):
+    '''
+    Parses and returns the scope and class name in `fqn`.
+    Example: if `fqn` is "NS1::NS2::C", "NS1::NS2" and "C" are returned.
+
+    Args:
+        fqn (string): fully-qualified class name.
+
+    Returns:
+        tuple[string,string]: tuple with the scope of and class name of `fqn`.
+    '''
+
+    pos = _find_namespace_end(fqn)
```

Comment:
I think this call could be substituted by `string.rfind`. I tried it with following examples:
```python
>>> a = "myclass"
>>> b = "ns::myclass"
>>> c = "ns1::ns2::myclass"
>>> d = "ROOT::RVec<float>"
>>> def _find_namespace_end(fqn):
...     last_found = -1
...     prev_c = ''
...     pos = 0
...     for c in fqn:
...         if c == ':' and prev_c == ':':
...             last_found = pos - 1
...         elif c == '<':
...             # If we found a template, this is already the class name,
...             # so we're done!
...             break
...         prev_c = c
...         pos += 1
...     return last_found
>>> _find_namespace_end(a)
-1
>>> _find_namespace_end(b)
2
>>> _find_namespace_end(c)
8
>>> _find_namespace_end(d)
4
>>> a.rfind("::")
-1
>>> b.rfind("::")
2
>>> c.rfind("::")
8
>>> d.rfind("::")
4
```

Reply:
Thanks for the comment, indeed `rfind` was my first implementation of this. Unfortunately, `rfind` alone does not work in the case where you have a templated type instantiated with a type in a namespace, e.g: `'ROOT::RDF::RInterface<ROOT::Detail::RDF::RLoopManager,void>'` . In that case, `rfind` would just hit the `::` in `ROOT::Detail::RDF::RLoopManager`, which is not what you want.

Reply:
Ah yes makes sense, I was sure there must have been some special case for this. Thanks for the clarification